README.md
# fernet

**This is forked from [fernet.js](https://github.com/csquared/fernet.js) and has been modified for better es6 support**

TypeScript implementation of <a href="https://github.com/kr/fernet-spec">Fernet symmetric encryption</a>.

Fernet is an opinionated way of using AES and HMAC authentication that makes
shared-secret symmetric encryption simpler for communicating applications.

Instead of using TypedArrays I use Hex Strings and CryptoJS's `Hex.parse`
to build up `CryptoJs.lib.WordArray` objects.

## WARNING

[It's generally _never_ considered safe to encrypt data in the browser.](http://www.matasano.com/articles/javascript-cryptography/)

However, you can use this library to encrypt/decrypt data server-side and decrypt data on a client.

That being said, the only randomness used by this library without your control is a call to `crypto.randomBytes` to generate IVs.
This function defaults to OpenSSL server-side and [browserify's random number generator implementation](https://github.com/crypto-browserify/crypto-browserify/blob/master/index.js)
client-side. The browser implementation only uses real browser crypto or throws an error. (IE: no calls to `Math.random()`)

If you're planning on generating the secrets in the browser do yourself a favor and get an audit.

## Usage

### Use in Vite

Important! if you are using this library in [vite](https://vitejs.dev/) projects, you will need to shim the `global` variable. This is necessary due to the usage of the builtin `crypto` library and native `Buffer` objects. To work around this, simply define `global` in the `vite.config.ts` file:

```ts
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  define: {
    global: 'window'
  },
  plugins: [vue()],
})
```

There is an included [demo](https://github.com/CalebM1987/fernet/tree/main/demo) application demonstrating the usage of this library. 

> note: See [Issue #1](https://github.com/CalebM1987/fernet/issues/1) for more details.
>
> ![app image](./demo/images/demo_app.png)

### node.js (use `Token` and `Secret` directly)

```js
import { Token, Secret } from "fernet";
```

## Using Fernet in `es6`

### set top level properties by modifying the `defaults` object.

```js
import { defaults, Secret } from "fernet";

// set secret
defaults.secret = new Secret("cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4=");
```

Or can use the `setSecret` function to modify the global `defaults`:

```js
import { setSecret } from "fernet";

setSecret("cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4=");
```

Sets the `secret` at the top level (`defaults` object) for all further Tokens made.

```js
import { defaults } from "fernet";

defaults.ttl = seconds; //seconds is number of seconds
```

Sets the `ttl` at the top level (`defaults` object) for all further Tokens made.

## Secret

### Generating a secret

    Generating appropriate secrets is beyond the scope of `Fernet`, but you should
    generate it using `/dev/random` in a *nix. To generate a base64-encoded 256 bit
    (32 byte) random sequence, try:

    dd if=/dev/urandom bs=32 count=1 2>/dev/null | openssl base64

### new Secret(string)

```javascript
import { Secret } from "fernet";

const secret = new Secret("cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4=");
/*
  {
    signingKeyHex: '730ff4c7af3d46923e8ed451ee813c87',
    signingKey: [CryptoJS.lib.WordArray],
    encryptionKeyHex: 'f790b0a226bc96a92de49b5e9c05e1ee',
    encryptionKey: [CryptoJS.lib.WordArray]
  }
*/
```

## Token

## new Token(options)

Options:

- `secret`: a `Secret` object
- `token`: a Fernet-encoded String
- `ttl`: seconds of ttl

For testing:

- `time`: Date object
- `iv`: Array of Integers

### full Token.encode example

```javascript
import { Token, Secret } from "fernet";

// before creating a token, we must have a Secret()
const secret = new Secret("cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4=");

//Have to include time and iv to make it deterministic.
//Normally time would default to (new Date()) and iv to something random.
const token = new Token({
  secret: secret
});
token.encode("Message"); // returns the encypted message
/*
'gAAAAABSO_yhAAECAwQFBgcICQoLDA0OD1PGoFV6wgWZG6AOBfQqevwJT2qKtCZ0EjKy1_TvyxTseR_3ebIF6Ph-xa2QT_tEvg=='
*/
```

### Token.decode example

Include tt

```js
import { Token, Secret } from "fernet";

// must use same secret that encoded the token
const secret = new Secret("cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4=");

const token = new Token({
  secret: secret,
  token:
    "gAAAAABSO_yhAAECAwQFBgcICQoLDA0OD1PGoFV6wgWZG6AOBfQqevwJT2qKtCZ0EjKy1_TvyxTseR_3ebIF6Ph-xa2QT_tEvg==",
  ttl: 0
});
token.decode();
```

### fernet.setSecret(string)

Sets the `secret` at the top level for all further Tokens made
from this instance of Fernet.

### defaults.ttl = seconds

Sets the `ttl` at the default for all further Tokens made.

## Build

to build, make sure you have [typescript](https://www.npmjs.com/package/typescript) and [ts-node](https://www.npmjs.com/package/ts-node) installed globally and run:

    > npm run build

## Test

    > npm test

tests node lib with [Jest](https://jestjs.io).




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.15](https://github.com/CalebM1987/fernet/compare/v0.0.14...v0.0.15) (2023-01-24)


### Features

* **demo:** improved demo with TTL support ([498e5ca](https://github.com/CalebM1987/fernet/commit/498e5ca24de958a4a1ea43abeed4534bae325b8b))

### [0.0.14](https://github.com/CalebM1987/fernet/compare/v0.0.13...v0.0.14) (2023-01-21)

### [0.0.13](https://github.com/CalebM1987/fernet/compare/v0.0.12...v0.0.13) (2023-01-21)


### Features

* added demo app and vite usage info ([a9edd6c](https://github.com/CalebM1987/fernet/commit/a9edd6cc06306ca03dbf5bc0d05fc3dc4e9cdc2f))

### [0.0.12](https://github.com/CalebM1987/fernet/compare/v0.0.11...v0.0.12) (2022-09-13)


### Bug Fixes

* **test:** use jsdom for node testing environment ([f31babb](https://github.com/CalebM1987/fernet/commit/f31babbd87243fe52e8dde1d6f6ef6b45ca86543))

### [0.0.11](https://github.com/CalebM1987/fernet/compare/v0.0.10...v0.0.11) (2022-09-13)

### [0.0.10](https://github.com/CalebM1987/fernet/compare/v0.0.9...v0.0.10) (2022-09-13)


### Bug Fixes

* try to polyfill global when it is not defined ([65a5c1f](https://github.com/CalebM1987/fernet/commit/65a5c1f95b268cbe527f8d28c219692f394959a5))

### [0.0.9](https://github.com/CalebM1987/fernet/compare/v0.0.8...v0.0.9) (2022-09-09)


### Bug Fixes

* fixed TTL validation bug ([332e36d](https://github.com/CalebM1987/fernet/commit/332e36d5f79df8946d029ddf02bcee9b51e3b7b1))

### [0.0.8](https://github.com/CalebM1987/fernet/compare/v0.0.7...v0.0.8) (2022-08-29)


### Features

* export necessary typings ([6781b80](https://github.com/CalebM1987/fernet/commit/6781b8088dcffa8aacdee099447954c4c3fd178d))

### [0.0.7](https://github.com/CalebM1987/fernet/compare/v0.0.6...v0.0.7) (2022-08-26)


### Features

* export all named exports ([e30e4d5](https://github.com/CalebM1987/fernet/commit/e30e4d5b9aec5dc9919c7fa95673cb280fd392de))

### [0.0.6](https://github.com/CalebM1987/fernet/compare/v0.0.5...v0.0.6) (2022-06-17)

### [0.0.5](https://github.com/CalebM1987/fernet/compare/v0.0.4...v0.0.5) (2022-06-16)

### [0.0.4](https://github.com/CalebM1987/fernet/compare/v0.0.3...v0.0.4) (2022-06-16)

### [0.0.3](https://github.com/CalebM1987/fernet/compare/v0.0.2...v0.0.3) (2022-06-16)

### 0.0.2 (2022-06-16)


### Features

* initial build of fernet package ([ae2445c](https://github.com/CalebM1987/fernet/commit/ae2445ca11957507cdbf56c7862d84f2cdfd0ed1))


src/lib/fernet.spec.ts
import { 
  Secret, 
  Token, 
  ArrayToHex, 
  createHmac, 
  timeBytes, 
  Hex, 
  hexBits, 
  urlsafe, 
  decode64toHex,
  Base64
} from './fernet';

// import fakeTimers from '@sinonjs/fake-timers'
import sinon from 'sinon'

const testData = {
  token:
    "gAAAAAAdwJ6wAAECAwQFBgcICQoLDA0ODy021cpGVWKZ_eEwCGM4BLLF_5CV9dOPmrhuVUPgJobwOz7JcbmrR64jVmpU4IwqDA==",
  now: "1985-10-26T01:20:00-07:00",
  iv: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
  src: "hello",
  secret: "cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4="
};

const unacceptableClockSkewTestData = {
  token:
    "gAAAAAAdwStRAAECAwQFBgcICQoLDA0OD3HkMATM5lFqGaerZ-fWPAnja1xKYyhd-Y6mSkTOyTGJmw2Xc2a6kBd-iX9b_qXQcw==",
  now: "1985-10-26T01:20:01-07:00",
  secret: "cw_0x689RpI-jtRR7oE8h_eQsKImvJapLeSbXpwF4e4="
};


const secret = new Secret(testData.secret);

describe('fernet.Token.prototype.decode', () => {

  const token = new Token({
    secret,
    token: testData.token
  })
  it('should decode token', () => {
    expect(token.decode()).toEqual('hello');
  });
  it('should cast to string', () => {
    expect(token.toString()).toEqual('hello');
  });
});

describe('version decoding test', ()=> {
  const token = new Token({
    secret,
    token: testData.token,
    version: 1
  })

  it('should recover version', ()=> {
    expect(token.version).toEqual(1)
    token.decode()
    expect(token.version).toEqual(128)
  })
})

describe('time decoding test', ()=> {
  const token = new Token({
    secret,
    token: testData.token,
    version: 1
  })

  it('should recover time', ()=> {
    token.decode()
    const now = new Date(Date.parse(testData.now))
    expect((token.time as Date).toUTCString()).toEqual(now.toUTCString())
  })
})

describe('IV decoding test', ()=> {
  const token = new Token({
    secret: secret,
    token: testData.token
  });
  it('should recover IV', ()=> {
    token.decode();
    const ivHex = ArrayToHex(testData.iv);
    expect(token.ivHex).toEqual(ivHex);
  })
})

describe('HMAC decoding test', ()=> {
  const token = new Token({
    secret: secret,
    token: testData.token
  });
  it('should recover HMAC', ()=> {
    token.decode()
    const computedHmac = createHmac(
      secret.signingKey,
      timeBytes(token.time as Date),
      token.iv!,
      token.cipherText!
    );
    expect(token.hmacHex).toBe(computedHmac.toString(Hex))
  })
})

describe('inherits parent TTL', ()=> {
  const token = new Token({
    secret,
    token: testData.token,
    ttl: 1
  })

  it('should throw invalid token TTL error', ()=> {
    expect(()=> token.decode()).toThrowError('Invalid Token: TTL')
  })

})

describe('throws wrong version error', ()=> {
  const tokenHex = decode64toHex(testData.token);
  const versionOffset = hexBits(8);
  const dirtyToken = "01" + tokenHex.slice(versionOffset);
  const tokenWords = Hex.parse(dirtyToken);
  const token = urlsafe(tokenWords.toString(Base64));
  const t = new Token({ secret: secret });
  it('raises new Error("Invalid version") on wrong version byte', ()=> {
    expect(()=> t.decode(token)).toThrowError("Invalid version");
  })
});

describe('validates HMAC', ()=> {
  const s = testData.token;
  const i = s.length - 5;
  const mutation = String.fromCharCode(s.charCodeAt(i) + 1);
  const dirtyHmacString = s.slice(0, i) + mutation + s.slice(i + 1);
  const token = new Token({
    secret: secret,
    token: dirtyHmacString
  });
  it('raises new Error("Invalid Token: HMAC") on wrong Hmac', ()=>{
    expect(()=> token.decode()).toThrowError('Invalid Token: HMAC')
  })
})

describe('validates far future timestamp', ()=> {

  it('raises new Error("far-future timestamp") on unacceptable clock skew', ()=> {
    const token = new Token({
      secret: new Secret(unacceptableClockSkewTestData.secret),
      token: unacceptableClockSkewTestData.token,
      ttl: 0
    });
    
    const clock = sinon.useFakeTimers(
      new Date(Date.parse(unacceptableClockSkewTestData.now)).getTime()
    )
    expect(()=> token.decode()).toThrowError('far-future timestamp')
    clock.restore()
  })
})

// /*****************************    ENCODING     **************************************** */
describe('encoding tests', ()=> {
  const token = new Token({
    secret,
    iv: testData.iv,
    time: testData.now
  })
  

  it('should encode message', ()=> {
    const encoded = token.encode(testData.src)
    expect(testData.token).toEqual(token.toString())
  })

  const tokenWithMessage = new Token({
    secret,
    iv: testData.iv,
    time: testData.now,
    message: testData.src
  })

  it("token.encode() makes token.toString() return the token", ()=> {
    tokenWithMessage.encode()
    expect(testData.token).toEqual(token.toString())
  })

  it("encode() returns the token as a String", ()=> {
    expect(token.encode(testData.src)).toEqual(testData.token)
  })
})

describe('testing default token arguments', ()=> {
  const token = new Token({
    secret,
    time: testData.now
  })

  it("randomly generates IV if one is not passed in", ()=> {
    const tokenString = token.encode(testData.src)
    expect(tokenString).not.toEqual(testData.token)

    const tokenString2 = token.encode(testData.src)
    expect(tokenString).not.toEqual(tokenString2)
  })

  const tokenWithoutTime = new Token({ secret })
  it('time defaults to Date.now()', ()=> {
    const cipherText = token.encode('foo')
    const recovered = token.decode(cipherText)
    expect(recovered).toEqual('foo')
  })
})


describe('fernet Secret tests', ()=> {
  
  const signingKeyHex = '730ff4c7af3d46923e8ed451ee813c87';
  const encryptionKeyHex = 'f790b0a226bc96a92de49b5e9c05e1ee';

  it('secret.signingKeyHex', ()=> {
    expect(secret.signingKeyHex).toEqual(signingKeyHex)
  })

  it('secret.signingKey', ()=> {
    expect(JSON.stringify(secret.signingKey))
      .toEqual(JSON.stringify(Hex.parse(signingKeyHex)))
  })

  it('secret.encryptionKeyHex', ()=> {
    expect(secret.encryptionKeyHex).toEqual(encryptionKeyHex)
  })

  it('secret.encryptionKey', ()=> {
    expect(JSON.stringify(secret.encryptionKey))
      .toEqual(JSON.stringify(Hex.parse(encryptionKeyHex)))
  })
  
  it('raises "new Error(\'Secret must be 32 url-safe base64-encoded bytes.\')" on wrong secret', ()=> {
    expect(()=> new Secret('not a good secret')).toThrowError('Secret must be 32 url-safe base64-encoded bytes.')
  })
})

describe('decodes within TTL time frame and throws TTL error after expiration', ()=> {

  it('should validate TTL before decoding', ()=> {
    const clock = sinon.useFakeTimers(
      new Date(Date.parse(unacceptableClockSkewTestData.now)).getTime()
    )
    const token = new Token({ secret: secret });
    token.encode(testData.src)
    // go forward 15 seconds
    clock.tick(15000)
    const dec = new Token({ secret, ttl: 30})
    expect(dec.decode(token.token)).toEqual(testData.src)

    // now make sure it throws TTL error after TTL expires
    clock.tick(30000)
    expect(()=> dec.decode(token.token)).toThrowError('Invalid Token: TTL')
    clock.reset()
  })
})


rollup.config.js
import dts from 'rollup-plugin-dts'
import esbuild from 'rollup-plugin-esbuild'
console.log('rollup config running')
const pkg = require('./package.json')
const name = pkg.main.replace(/\.js$/, '')

const bundle = config => ({
    ...config,
    input: 'src/index.ts',
    external: id => !/^[./]/.test(id),
})

export default [
    bundle({
        // define: {
        //     __VERSION__: pkg.version
        // },
        plugins: [esbuild({
            optimizeDeps: {
                bundle: true
            }
        })],
        output: [
            {
                file: `${name}.js`,
                format: 'cjs',
                sourcemap: true,
            },
            {
                file: `${name}.mjs`,
                format: 'es',
                sourcemap: true,
            },
        ],
    }),
    bundle({
        plugins: [dts()],
        output: {
            file: `${name}.d.ts`,
            format: 'es',
        },
    }),
]

src/lib/fernet.ts

if (typeof (window as any).global === 'undefined') {
  (window as any).global = window;
}
/**
 * This an es6 port of the library found here:
 *    https://github.com/csquared/fernet.js
 *
 * This was ported due to issues with the global module scope and caused issues with rollup.js.
 */
// import URLBase64 from "urlsafe-base64"; // having issues finding buffer
import { Buffer } from 'buffer';
import randomBytes from "randomBytes";
import AES from "crypto-js/aes";
import Utf8 from "crypto-js/enc-utf8";
import Hex from "crypto-js/enc-hex";
import Base64 from "crypto-js/enc-base64";
import HmacSHA256 from "crypto-js/hmac-sha256";

const defaults: Partial<FernetOptions> = {
  ttl: 0,
  versionHex: '80',
  secret: undefined
}

const URLBase64 = {
  encode(buffer: Buffer) {
    return buffer.toString('base64')
      .replace(/\+/g, '-') // Convert '+' to '-'
      .replace(/\//g, '_') // Convert '/' to '_'
      .replace(/=+$/, ''); // Remove ending '='
  },

  decode(base64: string) {

    // Add removed at end '='
    base64 += Array(5 - base64.length % 4).join('=');
  
    base64 = base64
      .replace(/\-/g, '+') // Convert '-' to '+'
      .replace(/\_/g, '/'); // Convert '_' to '/'
  
    return Buffer.from(base64, 'base64');
  },

  validate(base64: string) {
    return /^[A-Za-z0-9\-_]+$/.test(base64);
  }
}

type TextKey = string | CryptoJS.lib.WordArray;

/**
 * left pad a string for some hex conversions (changed to function rather than messing with the string prototype)
 *
 * @param {String} str - input string for padding
 * @param {String} padString - pad string to prepend input string
 * @param {Number} length - length of output padded string
 * @return {String} the padded string
 */
function lpad(str: string, padString: string, length: number) {
  while (str.length < length) str = padString + str;
  return str;
}

/**
 * Makes a Base64 string a url-safe base64 slertring
 * @param {String} string - input string to make url-safe
 * @return {String} a url-safe base64 string
 */
function urlsafe(string: string) {
  return string.replace(/\+/g, "-").replace(/\//g, "_"); //.replace(/=+$/, '')
}

/**
 * parse a Hex string to an Int
 * @param {String} hexString - a hexlified string
 * @return {Number} resulting integer from input hex string
 */
const parseHex = (hexString: string) => {
  return parseInt("0x" + hexString);
};

/**
 * turn bits into number of chars in a hex string
 * @param {Number} bits - input bits to convert to hex
 * @return {Number} number of chars in hex string
 */
function hexBits(bits: number) {
  return (bits / 8) * 2;
}

/**
 * convert base64 string to hex string
 * @param {String} string - input base64 string to hex
 * @return {String} a hex string
 */
function decode64toHex(string: string) {
  const s = URLBase64.decode(string.replace(/=+$/, ""));
  return Buffer.from(s).toString("hex");
}

/**
 * convert array to hex string
 * @param {Number[]} array - iv array of integers
 * @return {String} a hex string
 */
function ArrayToHex(array: number[]) {
  let hex = "";
  for (const _byte in array) {
    hex += lpad(Number(_byte).toString(16), "0", 2);
  }
  return hex;
}

/**
 * Creates a random hex string
 * @param {Number} size - size of hex string
 * @return {String} a hex string
 */
function randomHex(size: number): string {
  // return crypto.randomBytes(128/8).toString('hex');
  return randomBytes(128 / 8).toString("hex");
}

/**
 * Will safely create an IV Array of integers
 * @param {Number[]} [iv_array=null] - array of numbers for IV array.  If none passed in, a random hex will be created
 * @return {String} a hex string
 */
function setIV(iv_array?: CryptoJS.lib.WordArray): string {
  return Array.isArray(iv_array) ? ArrayToHex(iv_array) : randomHex(128 / 8);
}

/**
 * convert Time object or now into WordArray
 * @param {Date} time - input Date object to convert to WordArray
 * @return {Number[]} a word array
 */
function timeBytes(time: Date | number): CryptoJS.lib.WordArray {
  if (time) {
    //@ts-ignore
    time = time / 1000;
  } else {
    //@ts-ignore
    time = Math.round(new Date() / 1000);
  }
  //@ts-ignore
  const hexTime = lpad(time.toString(16), "0", "16");
  return Hex.parse(hexTime);
}

/**
 * convenience function to create a new instance of a Secret() and sets it as the default for all future tokens
 * @param {String} secret64 - base64 encoded secret string
 * @return {Secret} a Secret
 */
function setSecret(secret64: string) {
  defaults.secret = new Secret(secret64);
  return defaults.secret;
}

/**
 * Encrypts a message using AES
 * @param {String} message - message to encrypt
 * @param {String} encryptionKey - encryption key for AES
 * @param {Number[]} iv - IV array
 * @return {String} encrypted message
 */
function encryptMessage(message: string, encryptionKey: TextKey, iv: CryptoJS.lib.WordArray) {
  const encrypted = AES.encrypt(message, encryptionKey, { iv });
  return encrypted.ciphertext;
}

/**
 * Decrypts an AES Encrypted message
 * @param {String} cipherText - the encrypted message
 * @param {String} encryptionKey - decryption key for AES
 * @param {Number[]} iv - IV array
 * @return {String} decrypted message
 */
function decryptMessage(cipherText: TextKey, encryptionKey: TextKey, iv: CryptoJS.lib.WordArray) {
  const encrypted = {
    ciphertext: cipherText,
    key: encryptionKey,
    iv
  };

  const decrypted = AES.decrypt(encrypted as any, encryptionKey, { iv });
  return decrypted.toString(Utf8);
}

/**
 * creates an encryption token
 * @param {Number[]} signingKey - signing key for encyrption
 * @param {Date} time - time stamp for verification
 * @param {Number[]} iv - IV Array
 * @param {String} cipherText - the cipher text
 * @return {String} the url safe encrypted string
 */
function createToken(signingKey: TextKey, time: CryptoJS.lib.WordArray, iv: CryptoJS.lib.WordArray, cipherText: TextKey) {
  const hmac = createHmac(signingKey, time, iv, cipherText);
  let tokenWords = Hex.parse(defaults.versionHex!);
  for (let c of [time, iv, cipherText, hmac]) {
    tokenWords = tokenWords.concat(c as any);
  }
  return urlsafe(tokenWords.toString(Base64));
}

/**
 * Creates a SHA256 undigested byte string
 * @param {Number[]} signingKey - signing key for encyrption
 * @param {Date} time - time stamp for verification
 * @param {Number[]} iv - IV Array
 * @param {String} cipherText - the cipher text
 * @return {String} an undigested byte string
 */
function createHmac(signingKey: TextKey, time: CryptoJS.lib.WordArray, iv: CryptoJS.lib.WordArray, cipherText: TextKey) {
  let hmacWords = Hex.parse(defaults.versionHex!);
  for (let c of [time, iv, cipherText]) {
    hmacWords = hmacWords.concat(c as any);
  }
  return HmacSHA256(hmacWords, signingKey);
}

/**
 * Instance of a Secret to be used for the token encryption
 */
class Secret {
  public signingKeyHex: string;
  public signingKey: CryptoJS.lib.WordArray;
  public encryptionKeyHex: string;
  public encryptionKey: CryptoJS.lib.WordArray;
  /**
   * Creates a Secret to be used for the token encryption
   * @param {String} secret64 - base64 encoded secret string
   */
  constructor(secret64: string) {
    const secret = decode64toHex(secret64);
    if (secret.length !== hexBits(256)) {
      throw new Error("Secret must be 32 url-safe base64-encoded bytes.");
    }
    this.signingKeyHex = secret.slice(0, hexBits(128));
    this.signingKey = Hex.parse(this.signingKeyHex);
    this.encryptionKeyHex = secret.slice(hexBits(128));
    this.encryptionKey = Hex.parse(this.encryptionKeyHex);
  }
}

interface FernetOptions {
  /** time to live in seconds, if specfied a token 
   * whose lifespan is longer than TTL will not be decrypted 
   * */
  ttl?: number;
  /** the hex version number */
  versionHex?: string;
  /** the fernet Secret */
  secret?: Secret;
  /** message to encrypt */
  message: string;
  /** cipher text to decrypt */
  cipherText?: CryptoJS.lib.WordArray;
  /** token string */
  token?: string;
  /** version of token */
  version?: number;
  /** the IV array */
  iv?: CryptoJS.lib.WordArray | number[];
  /** date string */
  time: string ;
}

/**
 * Options for token object to perform encryption
 * @typedef TokenOptions
 * @property {Number} [ttl=60] - time to live in seconds
 * @property {Secret} secret - Secret object to use for encryption/decryption
 * @property {String} message - message to encrypt
 * @property {String} cipherText - cipher text to decrypt
 * @property {String} token - a token string
 * @property {String} [version='80'] - version of the token
 * @property {Number[]} iv - IV Array
 */

/**
 * Token object to perform encryption/decryption
 */
class Token {
  public secret?: Secret;
  public ttl?: number;
  public message: string;
  /** cipher text to decrypt */
  public cipherText?: CryptoJS.lib.WordArray;
  public cipherTextHex?: string;
  /** token string */
  public token?: string;
  /** version of token */
  public version?: string | number;
  /** the IV array */
  public optsIV?: CryptoJS.lib.WordArray;
  public maxClockSkew = 60;
  public time: Date | CryptoJS.lib.WordArray;
  public encoded: boolean;
  public iv?: CryptoJS.lib.WordArray;
  public ivHex?: string;
  public hmacHex?: string;

  /**
   * Token object to perform encryption/decryption
   * @param {TokenOptions} opts - options for token initialization
   */
  constructor(opts?: Partial<FernetOptions>) {
    opts = opts || {};
    this.secret = opts.secret || defaults.secret;
    this.ttl = opts.ttl || defaults.ttl;
    if (opts.ttl === 0) this.ttl = 0;
    this.message = opts.message ?? '';
    this.cipherText = opts.cipherText;
    this.token = opts.token ?? '';
    this.version = opts.version || parseHex(defaults.versionHex!);
    this.optsIV = opts.iv as CryptoJS.lib.WordArray;
    // @ts-ignore
    this.time = opts.time ? timeBytes(Date.parse(opts.time)) : timeBytes();
    this.encoded = false
  }

  /**
   * converts token to string
   * @return {String} to stringified token
   */
  toString() {
    return this.encoded ? this.token : this.message;
  }

  /**
   * Encrypts a message
   * @param {String} message - message to encrypt
   * @return {String} encoded token string
   */
  encode(message?: string) {
    if (!this.secret) throw new Error("Secret not set");
    this.encoded = true;
    this.ivHex = setIV(this.optsIV);
    this.iv = Hex.parse(this.ivHex); //if null will always be a fresh IV
    this.message = message || this.message;
    this.cipherText = encryptMessage(
      this.message,
      this.secret.encryptionKey,
      this.iv
    );
    this.token = createToken(
      this.secret.signingKey,
      this.time as CryptoJS.lib.WordArray,
      this.iv,
      this.cipherText
    );
    return this.token;
  }

  /**
   * Decrypts a token
   * @param {String} token - token to decrypt
   * @return {String} decoded message
   */
  decode(token?: string) {
    if (!this.secret) throw new Error("Secret not set");
    this.encoded = false;
    this.token = token || this.token;
    const tokenString = decode64toHex(this.token!);
    const versionOffset = hexBits(8);
    const timeOffset = versionOffset + hexBits(64);
    const ivOffset = timeOffset + hexBits(128);
    const hmacOffset = tokenString.length - hexBits(256);
    const timeInt = parseHex(tokenString.slice(versionOffset, timeOffset));
    this.version = parseHex(tokenString.slice(0, versionOffset));

    if (this.version != 128) {
      throw new Error("Invalid version");
    }

    this.time = new Date(timeInt * 1000);
    const currentTime = new Date()
    // @ts-ignore
    const timeDiff = (currentTime - this.time) / 1000

    if (this.ttl! > 0 && timeDiff > this.ttl!) {
      throw new Error("Invalid Token: TTL");
    }
    // @ts-ignore
    if (currentTime / 1000 + this.maxClockSkew < timeInt) {
      throw new Error("far-future timestamp");
    }

    this.ivHex = tokenString.slice(timeOffset, ivOffset);
    this.iv = Hex.parse(this.ivHex);
    this.cipherTextHex = tokenString.slice(ivOffset, hmacOffset);
    this.cipherText = Hex.parse(this.cipherTextHex);
    this.hmacHex = tokenString.slice(hmacOffset);
    const decodedHmac = createHmac(
      this.secret.signingKey,
      timeBytes(this.time),
      this.iv,
      this.cipherText
    );
    const decodedHmacHex = decodedHmac.toString(Hex);

    let accum = 0;
    for (let i = 0; i < 64; i++) {
      accum += decodedHmacHex.charCodeAt(i) ^ this.hmacHex.charCodeAt(i);
    }
    if (accum != 0) throw new Error("Invalid Token: HMAC");

    this.message = decryptMessage(
      this.cipherText,
      this.secret.encryptionKey,
      this.iv
    );
    return this.message;
  }
}

export {
  defaults,
  // buffer,
  Buffer,
  Secret,
  Token,
  TextKey,
  FernetOptions,
  setSecret,
  ArrayToHex,
  timeBytes,
  decode64toHex,
  createHmac,
  hexBits,
  urlsafe,
  Base64,
  Hex
};

